classes/security/Validation.inc.php

```diff
@@ -257,7 +311,33 @@ function generatePasswordResetHash($userId) {
 			// No such user
 			return false;
 		}
-		return substr(md5($user->getId() . $user->getUsername() . $user->getPassword()), 0, 6);
+		// create hash payload
+		$salt = Config::getVar('security', 'salt');
+		
+		// use last login time to create a basic expiry coefficient
+		$lastLogin = strtotime($user->getDateLastLogin());
+		$time = time();
+		$normalised = ($time - ($time % 3600)) + 7200;
+		$diff = floor(($normalised - $lastLogin) / 3600);
```

Comment:
Won't this restrict reset hashes to only working within the same hour-of-the-day?  That is, if I reset my password in the 59th minute, my hash would expire in the next 60 seconds.
```
$lastLogin = 02:30:55
$time = 05:59:00
$normalized = (05:59:00 - :59:00) + 02:00:00 = 07:00:00
$diff = 07:00:00 - 02:30:55 \ 01:00:00 = 4

$lastLogin = 02:30:55
$time = 06:01:00
$normalized = (06:01:00 - :01:00) + 02:00:00 = 08:00:00
$diff = 08:00:00 - 02:30:55 \ 01:00:00 = 5
```

For clarity, it might be helpful to define 3600 and 7200 in terms of a constant PASSWORD_RESET_SECONDS_UNTIL_EXPIRY or similar.